text/0000-pbft-consensus.md

```diff
@@ -0,0 +1,480 @@
+
++ `prepare`: Sent from every node to every other node; used as verification of
+the pre-prepare message.
+
++ `commit`: Sent from every node to every other node; used to determine if there
+is consensus that we should indeed commit the block contained in the original
+message.
+
++ `commit_final`: Sent from all nodes back to the primary node; used as
+confirmation that the block should be committed.
+
++ `checkpoint`: Sent by any node that has completed `checkpoint_period`
+`commit_final` messages.
+
++ `view_change`: Sent by any node that suspects that the primary node is faulty.
+
++ `new_view`: Sent by any node that receives `2f` `view_change` messages.
+
+#### Initialization
+At the beginning of the Engine's `start` method, some initial setup is
+required:
++ Create the node for processing messages
++ Establish timers and counters for checkpoint periods and block durations
++ Initialize the first working block
+
+#### Event Loop
+In a normal context (when the primary node is not faulty), the PBFT consensus
+algorithm operates as follows, inside the event loop of the `start` method:
+
+1. Receive a `BlockNew` message from the validator, representative of a
+   client's request. The primary node checks the legitimacy of the message and
+   assigns this message a sequence number, then broadcasts a `pre_prepare`
+   message to all nodes. Legitimacy is checked by looking at the `signer_id`
+   of the block in the `BlockNew` message, and making sure the `previous_id`
+   is valid as the current chain head. Secondary nodes ignore `BlockNew`
+   messages; only append them to their logs.
+
+2. Receive `pre_prepare` messages and check their legitimacy. `pre_prepare`
+   messages are legitimate if:
+  + `signer_id` and `summary` of block inside `pre_prepare` match the
+    corresponding fields of the original `BlockNew` block &&
+  + View in `pre_prepare` message corresponds to this server's current view &&
+  + This message hasn't been accepted already with a different `summary` &&
+  + Sequence number is within the sequential bounds of the log (low and high
+    water marks)
+
+   Once the `pre_prepare` is accepted, broadcast a `prepare` message to all nodes.
+
+3. Receive `prepare` messages, and check them all against their associated
+   `pre_prepare` message in this node's message log.
+
+4. Once the predicate `prepared` is true for this node, then call
+   `check_block`. If an error occurs (`ReceiveError` or `UnknownBlock`),
+   abort (call `cancel_block`). Otherwise, wait for a response (`BlockValid`
+   or `BlockInvalid`) from the validator. If `BlockValid`, then broadcast a
+   `commit` message to all other nodes. If `BlockInvalid`, then call
+   `cancel_block`.
+
+5. When the predicate `committed` is true for this node, then send a
+   `commit_final` message back to the primary.
+
+6. Once the primary has accumulated `f + 1` `commit_final` messages, then
+   it should commit the block using `commit_block`, and advance the chain
+   head.
+
```

Comment:
I believe both the prepare and commit phases require (2f+1) messages to be received, thus if f nodes are faulty, f+1 are not.

Reply:
Prepare and Commit phases do indeed require `2f + 1` messages. The `commit_final` messages were adapted from `reply` messages in the original paper, of which the "client" only needs `f + 1`. Do you think this is a reasonable adaptation from the original into the context of blockchain?